File: src/xt/lang/common_iter.clj
(ns xt.lang.common-iter
  (:require [std.lang :as l :refer [defspec.xt]])
  (:refer-clojure :exclude [constantly iterate repeatedly cycle range drop peek take map
                            mapcat concat filter keep partition take-nth]))

(l/script :xtalk
  {:require [[xt.lang.common-spec :as xt]]})

;;
;; ITERATOR
;;

(defn.xt iter-eq
  "checks that two iterators are equal"
  {:added "4.0"}
  [it0 it1 eq-fn]
  (xt/x:iter-eq it0 it1 eq-fn))

(defgen.xt iter-null
  "creates a null iterator"
  {:added "4.0"}
  []
  (xt/x:iter-null))

(defn.xt iter?
  "checks object is an iter"
  {:added "4.0"}
  [x]
  (return (or (xt/x:iter-native? x)
              (and (xt/x:is-object? x)
                   (xt/x:has-key? x
                                  "::"
                                  "iterator")))))

(defn.xt iter
  "converts to an iterator"
  {:added "4.0"}
  [x]
  (cond (xt/x:nil? x)
        (return (-/iter-null))
        
        (-/iter? x)
        (return x)

        (xt/x:iter-has? x)
        (return (xt/x:iter-from x))

        (xt/x:is-array? x)
        (return (xt/x:iter-from-arr x))
        
        (xt/x:is-object? x)
        (return (xt/x:iter-from-obj x))

        :else
        (return nil)))

(defn.xt collect
  "collects an iterator"
  {:added "4.0"}
  ([it f init]
   (var out init)
   (xt/for:iter [e it]
             (:= out (f out e)))
   (return out)))

(defn.xt nil<
  "consumes an iterator, returns nil"
  {:added "4.0"}
  [it]
  (xt/for:iter [e it])
  (return nil))

(defn.xt arr<
  "converts an array to iterator"
  {:added "4.0"}
  [it]
  (var out [])
  (xt/for:iter [e it]
    (xt/x:arr-push out e))
  (return out))

(defn.xt obj<
  "converts an array to object"
  {:added "4.0"}
  [it]
  (var out {})
  (xt/for:iter [e it]
    (xt/x:set-key out
                  (xt/x:first e)
                  (xt/x:second e)))
  (return out))

;;
;; ITER
;;

(defgen.xt constantly
  "constantly outputs the same value"
  {:added "4.0"}
  [val]
  (while true
    (yield val)))

(defgen.xt iterate
  "iterates a function and a starting value"
  {:added "4.0"}
  [f val]
  (while true
    (yield val)
    (:= val (f val))))

(defgen.xt repeatedly
  "repeatedly calls a function"
  {:added "4.0"}
  [f]
  (while true
    (yield (f))))

(defgen.xt cycle
  "cycles a function"
  {:added "4.0"}
  [seq]
  (var arr (:? (xt/x:is-array? seq) seq (-/arr< seq)))
  (if (== 0 (xt/x:len arr))
    (xt/x:err "Cannot be empty"))
  (while true
    (xt/for:array [e arr]
      (yield e))))

(defgen.xt range
  "setup a range function"
  {:added "4.0"}
  [x]
  (var arr    (:? (xt/x:is-array? x) x [x]))
  (var arrlen (xt/x:len arr))
  (var start  (:? (< 1 arrlen) (xt/x:first arr) 0))
  (var finish (:? (< 1 arrlen) (xt/x:second arr) (xt/x:first arr)))
  (var step   (:? (< 2 arrlen) (xt/x:get-idx arr 2) 1))
  (var i start)
  (cond (and (> step 0)
             (< start finish))
        (while (< i finish)
          (yield i)
          (:= i (+ i step)))
        
        (and (< step 0)
             (< finish start))
        (while (> i finish)
          (yield i)
          (:= i (+ i step)))

        :else (return)))

(defgen.xt drop
  "drop elements from seq"
  {:added "4.0"}
  ([n seq]
   (var i n)
   (xt/for:iter [e (-/iter seq)]
     (if (< 0 i)
       (:= i (- i 1))
       (yield e)))))

(defgen.xt peek
  "peeks at value and passes it on"
  {:added "4.0"}
  ([f seq]
   (xt/for:iter [e (-/iter seq)]
     (f e)
     (yield e))))

(defgen.xt take
  "take elements from seq"
  {:added "4.0"}
  ([n seq]
   (var i 0)
   (xt/for:iter [e (-/iter seq)]
     (if (< i n)
       (do (:= i (+ i 1))
           (yield e))
       (return)))))

(defgen.xt map
  "maps a function across seq"
  {:added "4.0"}
  ([f seq]
   (xt/for:iter [e (-/iter seq)]
     (yield (f e)))))

(defgen.xt mapcat
  "maps a function a concats"
  {:added "4.0"}
  ([f seq]
   (xt/for:iter [e0 (-/iter seq)]
     (var s0 (f e0))
     (xt/for:iter [e1 (-/iter s0)]
       (yield e1)))))

(defgen.xt concat
  "concats seqs into iterator"
  {:added "4.0"}
  ([seq]
   (xt/for:iter [e (-/mapcat (fn:> [x] x) seq)]
     (yield e))))

(defgen.xt filter
  "filters a seq using a function"
  {:added "4.0"}
  ([pred seq]
   (xt/for:iter [e (-/iter seq)]
     (if (pred e)
       (yield e)))))

(defgen.xt keep
  "keeps a seq using a function"
  {:added "4.0"}
  ([f seq]
   (xt/for:iter [e (-/iter seq)]
     (var v (f e))
     (if v (yield v)))))

(defgen.xt partition
  "partition seq into n items"
  {:added "4.0"}
  ([n seq]
   (if (> 1 n)
     (xt/x:err "Partition should be positive"))
   (var out [])
   (xt/for:iter [e (-/iter seq)]
     (if (< (xt/x:len out) n)
       (xt/x:arr-push out e)
       (do (yield out)
           (:= out []))))
   (if (< 1 (xt/x:len out))
     (yield out))))

(defgen.xt take-nth
  "takes first and then every nth item of a seq"
  {:added "4.0"}
  ([n seq]
   (if (> 1 n)
     (xt/x:err "Partition should be positive"))
   (var i 0)
   (xt/for:iter [e (-/iter seq)]
     (if (== i 0)
       (do (yield e)
           (:= i (- n 1)))
       (:= i (- i 1))))))

File: packages-gen/dart/libs/xt-lang/lib/common-iter.dart
iter_eq(it0, it1, eq_fn) {
  return (() {
    while(it0.moveNext()){
      if(!it1.moveNext()){
        return false;
      }
      if(!eq_fn(it0.current,it1.current)){
        return false;
      }
    }
    return !it1.moveNext();
  })();
}

iter_null() {
  return <dynamic>[].iterator;
}

iterp(x) {
  return (null != x) && (((null != x) && (x.runtimeType).toString().contains("Iterator")) || ((("Map" == (x.runtimeType).toString()) || (x.runtimeType).toString().startsWith("_Map") || (x.runtimeType).toString().startsWith("LinkedMap")) && (x.containsKey("::") && ("iterator" == x["::"]))));
}

iter(x) {
  if(null == x){
    return iter_null();
  }
  else if((() {
    var dart_truthy__50333 = iterp(x);
    return (null != dart_truthy__50333) && (false != dart_truthy__50333);
  })()){
    return x;
  }
  else if((x.runtimeType).toString().startsWith("List") || (x.runtimeType).toString().startsWith("_GrowableList")){
    return x.iterator;
  }
  else if(("Map" == (x.runtimeType).toString()) || (x.runtimeType).toString().startsWith("_Map") || (x.runtimeType).toString().startsWith("LinkedMap")){
    return x.entries.map((e) {
      return <dynamic>[e.key,e.value];
    }).iterator;
  }
  else if(((null != x) && (x.runtimeType).toString().contains("Iterator")) || (x.runtimeType).toString().contains("Iterable") || (x.runtimeType).toString().contains("List") || (x.runtimeType).toString().contains("Set")){
    return x.iterator;
  }
  else{
    return null;
  }
}

collect(it, f, init) {
  var out = init;
  var iter_50347 = iter(it);
  while(iter_50347.moveNext()){
    var e = iter_50347.current;
    out = f(out,e);
  };
  return out;
}

nil_lt(it) {
  var iter_50360 = iter(it);
  while(iter_50360.moveNext()){
    var e = iter_50360.current;
  };
  return null;
}

arr_lt(it) {
  var out = <dynamic>[];
  var iter_50373 = iter(it);
  while(iter_50373.moveNext()){
    var e = iter_50373.current;
    out.add(e);
  };
  return out;
}

obj_lt(it) {
  var out = <dynamic, dynamic>{};
  var iter_50386 = iter(it);
  while(iter_50386.moveNext()){
    var e = iter_50386.current;
    out[e[0]] = e[1];
  };
  return out;
}

constantly(val)  sync* {
  while(true){
    yield val;
  }
}

iterate(f, val)  sync* {
  while(true){
    yield val;
    val = f(val);
  }
}

repeatedly(f)  sync* {
  while(true){
    yield f();
  }
}

cycle(seq)  sync* {
  var arr = ((seq.runtimeType).toString().startsWith("List") || (seq.runtimeType).toString().startsWith("_GrowableList")) ? seq : arr_lt(seq);
  if(0 == arr.length){
    throw "Cannot be empty";
  }
  while(true){
    var arr_50399 = arr;
    for(var i50400 = 0; i50400 < arr_50399.length; ++i50400){
      var e = arr_50399[i50400];
      yield e;
    };
  }
}

range(x)  sync* {
  var arr = ((x.runtimeType).toString().startsWith("List") || (x.runtimeType).toString().startsWith("_GrowableList")) ? x : <dynamic>[x];
  var arrlen = arr.length;
  var start = (1 < arrlen) ? arr[0] : 0;
  var finish = (1 < arrlen) ? arr[1] : arr[0];
  var step = (2 < arrlen) ? arr[2] : 1;
  var i = start;
  if((step > 0) && (start < finish)){
    while(i < finish){
      yield i;
      i = (i + step);
    }
  }
  else if((step < 0) && (finish < start)){
    while(i > finish){
      yield i;
      i = (i + step);
    }
  }
  else{
    return;
  }
}

drop(n, seq)  sync* {
  var i = n;
  var iter_50421 = iter(seq);
  while(iter_50421.moveNext()){
    var e = iter_50421.current;
    if(0 < i){
      i = (i - 1);
    }
    else{
      yield e;
    }
  };
}

peek(f, seq)  sync* {
  var iter_50434 = iter(seq);
  while(iter_50434.moveNext()){
    var e = iter_50434.current;
    f(e);
    yield e;
  };
}

take(n, seq)  sync* {
  var i = 0;
  var iter_50447 = iter(seq);
  while(iter_50447.moveNext()){
    var e = iter_50447.current;
    if(i < n){
      i = (i + 1);
      yield e;
    }
    else{
      return;
    }
  };
}

map(f, seq)  sync* {
  var iter_50460 = iter(seq);
  while(iter_50460.moveNext()){
    var e = iter_50460.current;
    yield f(e);
  };
}

mapcat(f, seq)  sync* {
  var iter_50473 = iter(seq);
  while(iter_50473.moveNext()){
    var e0 = iter_50473.current;
    var s0 = f(e0);
    var iter_50486 = iter(s0);
    while(iter_50486.moveNext()){
      var e1 = iter_50486.current;
      yield e1;
    };
  };
}

concat(seq)  sync* {
  var iter_50499 = iter(mapcat((x) {
    return x;
  },seq));
  while(iter_50499.moveNext()){
    var e = iter_50499.current;
    yield e;
  };
}

filter(pred, seq)  sync* {
  var iter_50512 = iter(seq);
  while(iter_50512.moveNext()){
    var e = iter_50512.current;
    if((() {
      var dart_truthy__50334 = pred(e);
      return (null != dart_truthy__50334) && (false != dart_truthy__50334);
    })()){
      yield e;
    }
  };
}

keep(f, seq)  sync* {
  var iter_50525 = iter(seq);
  while(iter_50525.moveNext()){
    var e = iter_50525.current;
    var v = f(e);
    if(null != v){
      yield v;
    }
  };
}

partition(n, seq)  sync* {
  if(1 > n){
    throw "Partition should be positive";
  }
  var out = <dynamic>[];
  var iter_50538 = iter(seq);
  while(iter_50538.moveNext()){
    var e = iter_50538.current;
    if(out.length < n){
      out.add(e);
    }
    else{
      yield out;
      out = <dynamic>[];
    }
  };
  if(1 < out.length){
    yield out;
  }
}

take_nth(n, seq)  sync* {
  if(1 > n){
    throw "Partition should be positive";
  }
  var i = 0;
  var iter_50551 = iter(seq);
  while(iter_50551.moveNext()){
    var e = iter_50551.current;
    if(i == 0){
      yield e;
      i = (n - 1);
    }
    else{
      i = (i - 1);
    }
  };
}